### PR TITLE
docs: write a plugin guide

### LINKED ISSUE #22366: docs: provide `:help lua-plugin`
### Problem

Some users seem to find the documentation hard to navigate and have a hard time figuring out all the bits and pieces necessary to build Neovim plugins.

### Expected behavior

Neovim could provide a "write lua plugins" guide, which could contain a short description of how plugins are structured (plugin/, autoload/ etc) along with copy-pastable examples showing how to:

- create user-defined commands
- create autocommands
- spawn jobs 
- write plugin documentation
- make colorschemes?

It should not be exhaustive in order to remain easily consumed.

### PR DESCRIPTION
**Pull request by dundargoc**
This guide will aim to be in the same style as lua-guide, but aimed at new
plugin authors.

Closes https://github.com/neovim/neovim/issues/22366

Todo: check SIP

### CODE REVIEW

**Comment by dundargoc** [comment]
Heavily work in progress, as I need to write a plugin myself to determine which of this information is useful.

**Comment by dundargoc** [comment]
Very early stage at the moment, mostly gathering information and trying out these tips myself.

**Comment by echasnovski** [comment]
Here are some thoughts about framing "`setup()` vs. 'plugin/'" difference:

--------

Initialization side effects:

- Creating mappings.
- Creating default highlight groups.
- Creating autocommands inside a group which are needed for a correct work of plugin.
- Setting options.
- Setting variables (`g:`, `w:`, etc.).
- ... ?

Initialization side effects may be configurable:

- Whether to do it or not: to create mappings or not, to create default highlight groups or not, etc.
- Fine-tuning: which keys to use for mappings, which highlight groups to use as defaults, etc.

Where to put code for initialization side effects:

- If you (plugin author) need it to be executed during Neovim startup process and don't want them to be configurable, put it inside 'plugin/' directory. Make sure to document that it is needed that whole plugin is inside 'pack/*/start'.
- If you  need it to be executed for certain filetypes and don't want them to be configurable, put it inside 'ftplugin/' directory.
- Otherwise, put this code inside `setup()` (which takes `config` argument) in 'lua/\<plugin-name>.lua' file so that `require(<plugin-name>).setup()` is usable.

--------

Then teach how to properly write `setup()` so that it doesn't load all code base.

**Comment by clason** [comment]
Here's one point I'd like to see in the guide: How to structure your plugin to avoid performance issues (that then lead people to try to lazy-load it, although I don't really want to focus on **loading** plugins via plugin managers in this guide):

### Bad:

```lua
local foo = require'myplugin.foo'
local bar = require'myplugin.bar'

local M = {}

M.setup = function(opts)
  -- do stuff with opts
end

M.mybaz = foo.baz
M.myquux = bar.quux

return M
```

This will load and execute(!) the `myplugin/foo.lua` and `myplugin/bar.lua` when doing `require'myplugin'.setup()` or `local myplugin = require'myplugin'`.

### Good:

```lua
local foo = require'myplugin.foo'
local bar = require'myplugin.bar'

local M = {}

M.setup = function(opts)
  -- do stuff with opts
end

M.mybaz = function() require'myplugin.foo'.baz() end
M.mybaz = function() require'myplugin.bar'.quux() end

return M
```

This will _not_ load and execute these files when calling `require'myplugin'.setup()`, only when actually calling `require'myplugin'.mybaz()`.

The same holds for structuring `myplugin/foo.lua` and `myplugin/bar.lua`, of course.

**TL;DR:** Follow ORWYN, "only require when you need (it)".

**Comment by clason** [comment]
Other (adjacent) topics that could (should?) be addressed, at least briefly:
* writing documentation using docstrings and autogeneration
* release process (conventional commits, semantic versioning, automating releases)
* writing tests and setting up simple CI

Can't cover all the bases, and especially the automation part may have a too short half-life for static documentation, but a pointer on how to get started (and a nudge to get started at all) would be quite helpful.

I think @folke has a setup that is working pretty smoothly across his plugins?

File: runtime/doc/plugin-guide.txt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+                                       Type |gO| to see the table of contents.
+
+==============================================================================
+Introduction                                                      *plugin-guide*
+
+This guide will go through the basics of farming GitHub stars.
+
+Lua plugins are basically the same as a vim plugins, except the file extension
+is .lua instead of .vim and the file contains Lua code instead of vimscript.
```
Comment (clason) [inline runtime/doc/plugin-guide.txt:16]: Even though that sounds facetious, I think this is actually a very useful point to make up front: Lua plugins are first and foremost Neovim plugins, so (some of) the same rules apply.

If you add to that the difference that Lua plugins use `lua` and `require('foo').bar` instead of Vimscript's `autoload` and `call foo#bar`, you are basically full correct; everything else is either

1. due to _language_ (or API) differences, or
2. generic advice for _any_ Neovim plugin.

**Comment by max397574** [comment]
@clason ig you're talking about this
https://www.reddit.com/r/neovim/comments/103d9on/releases_docs_tests_for_plugins/

**Comment by clason** [comment]
> @clason ig you're talking about this
> https://www.reddit.com/r/neovim/comments/103d9on/releases_docs_tests_for_plugins/

No.

**Comment by hrsh7th** [comment]
I have a question.

I don't think it's common practice to delay require . I wonder if the official recommends this.

**Comment by justinmk** [comment]
Is a new file actually needed? From https://github.com/neovim/neovim/issues/22366#issuecomment-1445501976 I was imagining a section rather minimal section (2-3 pages long).

**Comment by ii14** [comment]
@hrsh7th yes, unfortunately it's not a common practice, and that's what we'd like to change. If you have top-level `require`s somewhere deep inside the plugin then it's fine, but the problem is when you start forcing users to unnecessarily load every single module just to change some plugin settings or just to make the plugin work. This might occasionally be needed, but most plugins do not need their entire implementation right off the bat.

And this leads to lazy loading in plugin managers, where often users have to manually redeclare every entry point to the plugin. A plugin knows best what is going to be needed when and it should be its responsibility to avoid doing any unnecessary work.

So the entry points should be as minimal as possible. If you need to define user commands, autocommands, key mappings - put them in `plugin/*.lua`, and make them call `require` *inside* the callback functions, so your plugin is not immediately loaded. If you need a `.setup{}` function, make `init.lua` do the least amount of work possible. If you need to have some public interface in `init.lua`, make it load the implementation inside the functions, just like clason pointed out.

**Comment by hrsh7th** [comment]
thank you. I understand.

It also makes sense to me to recommend lazy require in `plugin/*`.

**Comment by dundargoc** [comment]
> Nit: Is a new file actually needed? From #22366 (comment) I was imagining a section rather minimal section (2-3 pages long).

@justinmk I dunno, Lewis's comment had more likes so I went with that instead. Thumbs-up driven development. I don't mind a smaller guide section in an existing file. Were you thinking lua-guide?

**Comment by dundargoc** [comment]
I've read the comments here and participated in other discussions on this. I'm not currently convinced we have enough material to write a dedicated plugin guide.

Perhaps we can start by appending the material we currently have to `lua-guide`, and then separate it if we realize it becomes too large?

**Comment by dundargoc** [comment]
@justinmk we will see. I don't anticipate we have enough useful material for a new file.

**Comment by craigmac** [comment]
Could we reuse the material in usr_41.txt `:h write-plugin` in some way? It definitely would highlight how much easier it is to write the same plugin using Lua.

**Comment by dundargoc** [comment]
@craigmac I'll be honest: I'm not a big fan of `write-plugin`. It is very unfocused and some sections such as `using-<Plug>` have really confusing explanations. What parts exactly were you thinking we should reuse?

Also repetition bad and all that.

File: runtime/doc/plugin-guide.txt (hunk 1 above)
Comment (dundargoc) [reply runtime/doc/plugin-guide.txt:16]: Good point.

Comment (dundargoc) [reply runtime/doc/plugin-guide.txt:16]: That's actually a great point. We should mention that at the intro, but perhaps in a simplified manner. Thanks.

**Comment by craigmac** [comment]
@dundargoc I completely agree, every time I read it I realize I've forgotten half of the special stuff like `<Plug>`. I just thought we could link there to "how to do plugins like this in lua" so we'd cover similar things

**Comment by dundargoc** [comment]
> @dundargoc I completely agree, every time I read it I realize I've forgotten half of the special stuff like `<Plug>`. I just thought we could link there to "how to do plugins like this in lua" so we'd cover similar things

Hmm, I think I understand. I think we should mention `write-plugin` in the intro and make comparisons to vimscript (e.g. `require()` can be used as Lua equivalent to autoload, see @clason 's comment). Unsure exactly how this should look at the moment, but will write it in TODO. Thanks.

**Comment by echasnovski** [comment]
> What are people's thoughts on initialization methods? We can either use `plugin/` to automatically load the plugin or use `setup()` in `init.lua`. My impression is that the `plugin/` approach is superior as it's done automatically and doesn't require any user input. If anyone disagrees I'd like to know why and for which reasons.

I already weighed in on this:

> Where to put code for initialization side effects:
> 
>     * If you (plugin author) need it to be executed during Neovim startup process and don't want them to be configurable, put it inside 'plugin/' directory.
> 
>     * If you  need it to be executed for certain filetypes and don't want them to be configurable, put it inside 'ftplugin/' directory.
> 
>     * Otherwise, put this code inside `setup()` (which takes `config` argument) in 'lua/<plugin-name>.lua' file so that `require(<plugin-name>).setup()` is usable.

Downsides of using 'plugin/' come directly from its definition of "runs at startup":
- It **always** runs at startup meaning that full lazy-loading is not possible. Yes, only "cheap" code should be in 'plugin/', but this is still an issue.
- User configuration through Lua is trickier than with direct `setup()`. It should be done either with Vimscript `g:` variables or some other exported Lua mechanism (at which point why not use `setup()` directly?).

**Comment by clason** [comment]
> What are people's thoughts on initialization methods? We can either use `plugin/` to automatically load the plugin or use `setup()` in `init.lua`. My impression is that the `plugin/` approach is superior as it's done automatically and doesn't require any user input. If anyone disagrees I'd like to know why and for which reasons.

Neither is strictly superior; it is (or should be) a conscious design decision whether something (everything) should be done automatically on Neovim startup (-> put it in `plugin/`) or only explicitly by the user at a time of their choosing (-> put it in a `setup()`, or generally a `mymodule/init.lua`).

This is not in any way different from "legacy" vimscript plugins (just replace `setup()` by some `autoload/` function).

**Comment by dundargoc** [comment]
> Downsides of using 'plugin/' come directly from its definition of "runs at startup":
> - It **always** runs at startup meaning that full lazy-loading is not possible. Yes, only "cheap" code should be in 'plugin/', but this is still an issue.

I don't understand why this is a problem. Either the plugin is autoloaded at startup by being in `plugin/`, or the user starts the plugin by loading it themselves at startup. Either way the code is run at startup. Am I missing a scenario?

> - User configuration through Lua is trickier than with direct `setup()`. It should be done either with Vimscript `g:` variables or some other exported Lua mechanism (at which point why not use `setup()` directly?).

Yes, but this was about initialization only, not configuration. Unless you meant if you're already using configuration with `setup()` you migh as well use it for initialization as well?

**Comment by dundargoc** [comment]
> Neither is strictly superior; it is (or should be) a conscious design decision whether something (everything) should be done automatically on Neovim startup (-> put it in `plugin/`) or only explicitly by the user at a time of their choosing.

What time would this be other than startup? Do you have any real examples?

**Comment by clason** [comment]
>  Unless you meant if you're already using configuration with setup() you might as well use it for initialization as well?

That is a very good point that serves to clarify the discussion: configuration and initialization are different things, but often conflated. The `setup()` pattern in particular combines the two into a single step (hence the name `setup` instead of `config`) -- which can be the right thing _for some plugins_.

Again, the point is to be aware of the consequences of certain patterns so people can pick the option that is right for their plugin, rather than blindly copying patterns "because that's how Lua plugins are written".

**Comment by dundargoc** [comment]
> Again, the point is to be aware of the consequences of certain patterns so people can pick the option that is right for their plugin, rather than blindly copying patterns "because that's how Lua plugins are written".

And again, I am asking for real examples of when you'd want to choose one approach over the other. A "use the right tool for the right job" does not a guide make.

**Comment by dundargoc** [comment]
> The whole idea of "lazy loading". That is, not loading plugin during startup but after it (ideally just before it is needed) in order reduce startup time. Example: initialize and apply configuration of auto-completion engine on first `InsertEnter` event.

Ah, right. So that's what you meant by full lazy-loading isn't possible with `plugin` since some code *has* to run at startup. I guess the best case scenario is that the `plugin/` code is really cheap, and in the worse case it's really expensive and no work is being deferred. Hmm, this wasn't as simple and one-sided as I initially thought...

> Plugin author might want to make configurable the Initialization process itself. Example: allow user to disable creation of some particular autocommand for the feature user doesn't want.

Hmm, am I right that this wouldn't have been possible with the `plugin/` approach because the autocommand would've already been defined before the user has a chance to specify they don't want it?

**Comment by dundargoc** [comment]
> I prefer the plugin/ approach because when I install a plugin into a `pack/*/start` folder, I shouldn't need to run a setup method to have commands and keybindings setup. See Telescope source for how this is done correctly. It makes the :Telescope command available.

I agree with you and I think this feels the most "natural".

> plugin/ doesn't always load, it only loads if it's in `pack/*/start`, otherwise it loads when you use packadd specifically, so you could only load packages in your `after/ftplugin/` files.

Wait, maybe this is because I don't really use `packadd`, but doesn't a file under `plugin/` always load at startup regardless?

**Comment by craigmac** [comment]
Nothing in `pack/*/opt` is sourced until you explicitly use `packadd pkgname`

**Comment by dundargoc** [comment]
> Nothing in `pack/*/opt` is sourced until you explicitly use `packadd pkgname`

Yes, but I was specifically talking about `plugin/`. I feel like optional packages in `pack/` are out of scope for the guide.

File: runtime/doc/help.txt [hunk 2]
```diff
@@ -103,6 +103,7 @@ API (EXTENSIBILITY/SCRIPTING/PLUGINS)
 |api|			Nvim API via RPC, Lua and VimL
 |ui|			Nvim UI protocol
 |lua-guide|		Nvim Lua guide
+|plugin-guide|		Nvim plugin guide
```
Comment (dundargoc) [inline runtime/doc/help.txt:106]: Delete

Comment (dundargoc) [inline runtime/doc/help.txt:106]: Delete

**Review by dundargoc — COMMENTED** [review]
LGTM!

(How do I delete review comments?)

**Comment by duarm** [comment]
Sorry if this isn't the appropriate place, but the discussion about separating setup and config reminded me of a plugin I tried to make which had an UI for configuring plugins like you do in vscode, It ultimately didn't work because no plugin exposed it's options so I could easily get them. Separating the opts table into a config.lua or options.lua would allow me to require all the available options with ` require "*.config"`. I think this might be a option worth considering for the guide.

**Comment by dundargoc** [comment]
@duarm it's appropriate, no worries :).

The plugin guide writing has lost significant steam due to the fact there really isn't much consensus on what constitutes a good practice, even from veteran neovimmers. The only realistic way forward I see for the forseeable future is to significantly scale back the ambition level and write a short "quickstart" guide to get plugin authors up and running. The current PR would be the skeleton for what I had in mind.

**Comment by dundargoc** [comment]
I was significantly more optimistic when I started this than I am now. After extensive discussion I don't believe there's much "best practises" to go around; people seem mostly interested in propagating their preferred stylistic choices without much basis. Feel free to take over.

**Comment by duarm** [comment]
I think instead of a plugin guide, it should be a "plugin standard", instead of "best practices", it should demand the plugin to work in a particular way, without actually breaking it, it's just a standard . 
A general guide for plugins can be arbitrarily written just to provide a north for those who want to start, a well commented simple plugin + the built-in ones are enough.
